### PR TITLE
solution 18/08

### PR DESCRIPTION
**Pull request by VadimKadasiev**

### CODE REVIEW

File: src/main/java/book/store/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    @PostMapping
+    @PreAuthorize("hasRole('USER')")
+    public OrderResponseDto createOrder(Authentication authentication,
+                                        @RequestBody OrderRequestDto orderRequestDto) {
+        return orderService.createOrder(getCurrentUser(authentication), orderRequestDto);
+    }
+
+    @GetMapping
+    @PreAuthorize("hasRole('USER')")
+    public List<OrderResponseDto> getOrders(Authentication authentication, Pageable pageable) {
```
Comment (sarakhmen) [inline src/main/java/book/store/controller/OrderController.java]: please add default pagable params. fix in all places

File: src/main/java/book/store/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        return orderService.getOrderItems(id, getCurrentUser(authentication),pageable);
+    }
+
+    @PatchMapping("{id}")
+    @PreAuthorize("hasRole('ADMIN')")
+    public void changeOrderStatus(@PathVariable Long id, @RequestBody String status) {
+        status = status.replace("\"", "")
+                .replace("}", "")
+                .trim()
+                .substring(15);
```
Comment (sarakhmen) [inline src/main/java/book/store/controller/OrderController.java]: why did you modify status value here? it should be passed in a correct format

File: src/main/java/book/store/service/impl/OrderServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            order.setTotal(order.getTotal()
+                    .add(book.getPrice()
+                            .multiply(BigDecimal
+                                    .valueOf(cartItemResponseDto.getQuantity()))));
+            orderItemRepository.save(orderItem);
+            cartItemRepository.deleteById(cartItemResponseDto.getId());
+        }
+    }
+
+    public List<OrderResponseDto> getOrders(User user, Pageable pageable) {
```
Comment (sarakhmen) [inline src/main/java/book/store/service/impl/OrderServiceImpl.java:83]: add @override in all places where needed

File: src/main/java/book/store/service/impl/OrderServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    @Override
+    public Set<OrderItemResponseDto> getOrderItems(Long orderId, User user, Pageable pageable) {
+        checkUserOrderMatching(user, orderId);
+        return orderItemRepository.findOrderItemByOrderId(orderId).stream()
+                .map(orderItemMapper::toOrderItemResponseDto).collect(Collectors.toSet());
+    }
+
+    public void changeOrderStatus(Long orderId, Order.OrderStatus orderStatus) {
```
Comment (sarakhmen) [inline src/main/java/book/store/service/impl/OrderServiceImpl.java:111]: add transactional

**Review by sarakhmen — CHANGES_REQUESTED** [review]
let's slightly improve your solution :)

File: src/main/java/book/store/controller/OrderController.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/orders")
+public class OrderController {
+    private final OrderService orderService;
+
+    @PostMapping
+    @PreAuthorize("hasRole('USER')")
+    public OrderResponseDto createOrder(Authentication authentication,
+                                        @RequestBody OrderRequestDto orderRequestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/controller/OrderController.java]:
```suggestion
                                        @RequestBody @Valid OrderRequestDto orderRequestDto) {
```

File: src/main/java/book/store/model/Order.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Enumerated(EnumType.STRING)
+    @Column(nullable = false)
+    private OrderStatus status = OrderStatus.PENDING;
+
+    @Column(nullable = false)
+    private BigDecimal total = BigDecimal.ZERO;
+
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/Order.java]:
```suggestion
    private LocalDateTime orderDate = LocalDateTime.now();
```

File: src/main/java/book/store/model/Order.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+@Table(name = "orders")
+@SQLDelete(sql = "UPDATE order SET is_deleted = true WHERE id = ?")
+@SQLRestriction(value = "is_deleted = FALSE")
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/Order.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
```

File: src/main/java/book/store/model/Order.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    private BigDecimal total = BigDecimal.ZERO;
+
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
+
+    @Column(nullable = false)
+    private String shippingAddress;
+
+    @OneToMany
+    @JoinColumn(name = "order_id")
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/Order.java]:
```suggestion
    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/book/store/model/Order.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    private Set<OrderItem> orderItems;
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/Order.java:53]:
```suggestion
    private Set<OrderItem> orderItems;
    
    @Column(nullable = false)
    private boolean isDeleted;
```

File: src/main/java/book/store/model/OrderItem.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+@Getter
+@Setter
+@Table(name = "order_items")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/OrderItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "order_id", nullable = false)
```

File: src/main/java/book/store/service/impl/OrderServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        order.setOrderDate(LocalDateTime.now());
+        order.setShippingAddress(orderRequestDto.getShippingAddress());
+        order.setUser(user);
+        orderRepository.save(order);
+        createOrderItems(user, order);
+        order.setOrderItems(orderItemRepository.findOrderItemByOrderId(order.getId()));
+        return orderMapper.orderToOrderResponseDto(order);
+    }
+
+    @Transactional
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/OrderServiceImpl.java]: Transactional won't work here

File: src/main/java/book/store/service/impl/OrderServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    public void createOrderItems(User user, Order order) {
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/OrderServiceImpl.java]:
```suggestion
    private void createOrderItems(User user, Order order) {
```

File: src/main/java/book/store/service/impl/OrderServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+                orderMapper::orderToOrderResponseDto).toList();
+    }
+
+    @Override
+    public OrderItemResponseDto getOrderItem(Long orderId, long orderItemId, User user) {
+        checkUserOrderMatching(user, orderId);
+        return orderItemMapper.toOrderItemResponseDto(orderItemRepository
+                .findOrderItemByOrderIdAndId(orderId, orderItemId)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find order item by id "
+                        + orderItemId + " in order by ID " + orderId)));
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/OrderServiceImpl.java]:
```suggestion
        OrderItem item = orderItemRepository.findByIdAndOrderIdAndUserId(orderItemId, orderId, userId)
                .orElseThrow(() -> new EntityNotFoundException(
                        .....
                ));
        return orderItemMapper.toOrderItemDto(item);
```

File: src/main/java/book/store/service/impl/OrderServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    @Override
+    public Set<OrderItemResponseDto> getOrderItems(Long orderId, User user, Pageable pageable) {
+        checkUserOrderMatching(user, orderId);
+        return orderItemRepository.findOrderItemByOrderId(orderId).stream()
+                .map(orderItemMapper::toOrderItemResponseDto).collect(Collectors.toSet());
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/OrderServiceImpl.java]:
```suggestion
        Order order = orderRepository.findByIdAndUserId(orderId, userId)
                .orElseThrow(() -> new EntityNotFoundException(
                        .....
                ));
        return orderItemMapper.toOrderItemDtoList(order.getOrderItems());
```

File: src/main/java/book/store/service/impl/OrderServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+                                    .valueOf(cartItemResponseDto.getQuantity()))));
+            orderItemRepository.save(orderItem);
+            cartItemRepository.deleteById(cartItemResponseDto.getId());
+        }
+    }
+
+    @Override
+    public List<OrderResponseDto> getOrders(User user, Pageable pageable) {
+        return orderRepository.findAllByUserId(user.getId()).stream().map(
+                orderMapper::orderToOrderResponseDto).toList();
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/OrderServiceImpl.java]:
```suggestion
        return orderRepository.findAllByUserId(user.getId())
        .stream()
        .map(orderMapper::orderToOrderResponseDto)
        .toList();
```

File: src/main/java/book/store/service/impl/OrderServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+    @Override
+    @Transactional
+    public void changeOrderStatus(Long orderId, Order.OrderStatus orderStatus) {
+        Order order = orderRepository.findById(orderId).orElseThrow();
+        order.setStatus(orderStatus);
+        orderRepository.save(order);
+    }
+
+    private void checkUserOrderMatching(User user, Long orderId) {
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/OrderServiceImpl.java]: remove

File: src/main/java/book/store/service/impl/OrderServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+    public Set<OrderItemResponseDto> getOrderItems(Long orderId, User user, Pageable pageable) {
+        checkUserOrderMatching(user, orderId);
+        return orderItemRepository.findOrderItemByOrderId(orderId).stream()
+                .map(orderItemMapper::toOrderItemResponseDto).collect(Collectors.toSet());
+    }
+
+    @Override
+    @Transactional
+    public void changeOrderStatus(Long orderId, Order.OrderStatus orderStatus) {
+        Order order = orderRepository.findById(orderId).orElseThrow();
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/OrderServiceImpl.java]:
```suggestion
        Order order = orderRepository.findById(orderId).orElseThrow(() -> new EntityNotFoundException(....
```

File: src/main/java/book/store/service/impl/OrderServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+    private final OrderItemRepository orderItemRepository;
+    private final OrderItemMapper orderItemMapper;
+    private final CartItemRepository cartItemRepository;
+
+    @Override
+    @Transactional
+    public OrderResponseDto createOrder(User user, OrderRequestDto orderRequestDto) {
+        if (shoppingCartService.getShoppingCart(user.getId()).getCartItems().isEmpty()) {
+            throw new RuntimeException("Shopping Cart is empty");
+        }
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/OrderServiceImpl.java]:
```suggestion
        ShoppingCart cart = shoppingCartRepository.findByUserId(userId);
        if (cart.getCartItems().isEmpty()) {
            throw new OrderProcessingException("Cart is empty for user: " + userId);
        }
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**